lib/pure/times.nim

```diff
@@ -1396,228 +1488,721 @@ proc `*=`*[T: TimesMutableTypes, U](a: var T, b: U) =
-  ##    ss         Same as above but always two digits. 0 is prepended.                               ``00:00:06 -> 06``
-  ##    t          ``A`` when time is in the AM. ``P`` when time is in the PM.
-  ##    tt         Same as above, but ``AM`` and ``PM`` instead of ``A`` and ``P`` respectively.
-  ##    y(yyyy)    This displays the year to different digits. You most likely only want 2 or 4 'y's
-  ##    yy         Displays the year to two digits.                                                   ``2012 -> 12``
-  ##    yyyy       Displays the year to four digits.                                                  ``2012 -> 2012``
-  ##    z          Displays the timezone offset from UTC.                                             ``GMT+7 -> +7``, ``GMT-5 -> -5``
-  ##    zz         Same as above but with leading 0.                                                  ``GMT+7 -> +07``, ``GMT-5 -> -05``
-  ##    zzz        Same as above but with ``:mm`` where *mm* represents minutes.                      ``GMT+7 -> +07:00``, ``GMT-5 -> -05:00``
-  ##    fff        Milliseconds display                                                               ``1000000 nanoseconds -> 1``
-  ##    ffffff     Microseconds display                                                               ``1000000 nanoseconds -> 1000``
-  ##    fffffffff  Nanoseconds display                                                                ``1000000 nanoseconds -> 1000000``
-  ## ============  =================================================================================  ================================================
-  ##
-  ## Other strings can be inserted by putting them in ``''``. For example
-  ## ``hh'->'mm`` will give ``01->56``.  The following characters can be
-  ## inserted without quoting them: ``:`` ``-`` ``(`` ``)`` ``/`` ``[`` ``]``
-  ## ``,``. However you don't need to necessarily separate format specifiers, a
-  ## unambiguous format string like ``yyyyMMddhhmmss`` is valid too.
+type
+  AmPm = enum
+    apUnknown, apAm, apPm
+
+  Era = enum
+    eraUnknown, eraAd, eraBc
+
+  ParsedTime = object
+    amPm: AmPm
+    era: Era
+    year: Option[int]
+    month: Option[int]
+    monthday: Option[int]
+    utcOffset: Option[int]
+
+    # '0' as default for these work fine
+    # so no need for `Option`.
+    hour: int
+    minute: int
+    second: int
+    nanosecond: int
+
+  LayoutTokenKind = enum
+    tkPattern, tkLiteral
+
+  LayoutPattern {.pure.} = enum
+    d, dd, ddd, dddd
+    h, hh, H, HH
+    m, mm, M, MM, MMM, MMMM
+    s, ss
+    fff, ffffff, fffffffff
+    t, tt
+    y, yy, yyy, yyyy, yyyyy
+    YYYY
+    uuuu
+    UUUU
+    z, zz, zzz, zzzz
+    g
+
+    # This is a special value used to mark literal layout values.
+    # See the doc comment for ``TimeLayout.patterns``.
+    Lit
+
+  TimeLayout* = object ## Represents a format for parsing and printing
+                       ## time types.
+    patterns: seq[byte] ## \
```

Comment:
Why is this encoded as bytes? Wouldn't ``seq[LayoutPattern]`` make more sense?

Reply:
Also, I think ``LayoutPattern`` should be called ``TimePattern``. It doesn't have much to do with the layout of the pattern so I'm not sure why you named it this way.

Reply:
I now noticed that you are referring to these format specifiers as "layout patterns" which is just confusing to me. Layout to me means "add 5 spaces before this string" or "indent and wrap these two lines so that they fit 80 characters", it's not about formatting time.

Reply:
Please rename all of these types and use the word "Format" instead of "Layout"

Reply:
The reason I wanted to avoid using "format" is because of the ambiguity (since it can be both a verb and a noun in this context). But English isn't my first language and you're probably right that using "format" anyway is better :)

Reply:
> Why is this encoded as bytes? Wouldn't seq[LayoutPattern] make more sense?

See the doc comment for this field. Basically `TimeLayout.patterns` not only contains `LayoutPattern` values, but also arbitrary bytes that are treated as text. This is a bit hackish, but it seems to performs well.

Reply:
> Isn't this ambiguous? dddd.byte == 3.byte? What if I want \3 in my string?

Each literal sequence is prefixed by `LayoutPattern.Lit` and the length of the literal sequence